mlflow/tracking/fluent.py

```diff
@@ -219,8 +219,10 @@ def __enter__(self):
         return self
 
     def __exit__(self, exc_type, exc_val, exc_tb):
-        status = RunStatus.FINISHED if exc_type is None else RunStatus.FAILED
-        end_run(RunStatus.to_string(status))
+        active_run_stack = _active_run_stack.get()
+        if len(active_run_stack) > 0 and active_run_stack[-1] is self:
+            status = RunStatus.FINISHED if exc_type is None else RunStatus.FAILED
+            end_run(RunStatus.to_string(status))
```

Comment:
Should we always call `end_run`? The current code looks like there is a case where it's not called.

Reply:
I think no. the code is inside `run_context.__exit__`, it should only end the run of itself.

Reply:
Yes. The example in my PR description is the scenario:
```
import mlflow

mlflow.set_experiment("nested_run_debugging")
with mlflow.start_run():
    for i in range(10):
        with mlflow.start_run(nested=True):
            mlflow.log_param("index", i)
            if i == 3:
                mlflow.end_run("FAILED")
```

Reply:
It's not nested.

<img width="657" alt="image" src="https://github.com/user-attachments/assets/456fc0f4-b6d7-4009-a7f6-259b28a4a40e">

Reply:
ah I see, each thread has an isolated `active_run_stack`. Is this a bug?

Reply:
uh ? in my test the result is nested:
![Uploading image.png…]()

Reply:
You can run my `ThreadPoolExecutor` code.

Reply:
> Ah I see, each thread has an isolated active_run_stack

yes this is for thread-safety. We can't share the run_stack across different thread, otherwise the stack behavir is broken.
so that "is_nested" functionality can't be used in multi-threading cases.

in multi-threading case, you can set `parent_run_id` instead